corgi/core/models.py

```diff
@@ -327,13 +328,19 @@ class ProductModel(TimeStampedModel):
     ofuri = models.CharField(max_length=1024, default="")
     lifecycle_url = models.CharField(max_length=1024, default="")
 
-    # Override each of the below on that model, e.g. products = None on the Product model
-    pnodes = GenericRelation(ProductNode)  # Needed to avoid a mypy warning
-    products = fields.ArrayField(models.CharField(max_length=200), default=list)
-    product_versions = fields.ArrayField(models.CharField(max_length=200), default=list)
-    product_streams = fields.ArrayField(models.CharField(max_length=200), default=list)
-    product_variants = fields.ArrayField(models.CharField(max_length=200), default=list)
-    channels = fields.ArrayField(models.CharField(max_length=200), default=list)
+    pnodes = GenericRelation(ProductNode, related_query_name="%(class)s")
+    # Override below (e.g. products = None) on the Product model
+    products = models.ForeignKey("Product", on_delete=models.CASCADE, related_name="%(class)ss")
+    # Override below on the Product and ProductVersion models
+    productversions = models.ForeignKey(
+        "ProductVersion", on_delete=models.CASCADE, related_name="%(class)ss"
+    )
+
+    # below fields are added implicitly to all ProductModel subclasses
+    # using Django reverse relations, but mypy needs an explicit type hint
+    productstreams: Union[models.ForeignKey, models.Manager["ProductStream"]]
+    productvariants: Union[models.ForeignKey, models.Manager["ProductVariant"]]
+    channels: models.Manager["Channel"]
```

Comment:
I'm not a huge fan of these abstracted fields, it makes it harder to read the child models and they still contain tons of comments referring to these same fields. We could just define them there instead and keep this abstract model for the fields that are the same in all models?

Reply:
Yeah....a lot of this was "thinking on my feet", where I saw some duplication I tried to get rid of, instead of needing to change the same thing in 5 different places. Originally this was cleaner and more useful / reusable, but then mypy started complaining and I had to hack around that. I'll just put this back and stop trying to be clever.

But one caveat is we do need *something* defined on the ProductModel base class, so that mypy knows these properties exist and our other methods can stay generic (just handle a ProductModel). So the type hints will need to stay even after the fields themselves get moved to below,

Another caveat is that this gets complicated due to the ForeignKey relationship. We define e.g. a productstreams field on the ProductVariant model, and all ProductModel subclasses will have this productstreams field.

But the type is different depending on which direction of the relationship you're looking at, like I describe in the MR description. Following the ForeignKey link "up" from variants to streams gives you one parent object directly. Whereas following the ForeignKey link "down" from products or versions to streams gives you a set of many related child objects that you have to iterate over.

Using a ManyToManyField would simplify that approach and make this taxonomy identical to the others. Maybe there's reasons not to do so, though - would appreciate any thoughts you have.

Reply:
> But one caveat is we do need something defined on the ProductModel base class, so that mypy knows these properties exist and our other methods can stay generic (just handle a ProductModel). So the type hints will need to stay even after the fields themselves get moved to below,
> 

That's odd. I don't see other abstract models having to define their child fields just to keep mypy happy. Looking at e.g. the Tag model, even if I comment out the `tagged_model` abstract method, mypy doesn't complain about `Tag` not having the `tagged_model` field defined.

> Another caveat is that this gets complicated due to the ForeignKey relationship. We define e.g. a productstreams field on the ProductVariant model, and all ProductModel subclasses will have this productstreams field.
> 
> But the type is different depending on which direction of the relationship you're looking at, like I describe in the MR description. Following the ForeignKey link "up" from variants to streams gives you one parent object directly. Whereas following the ForeignKey link "down" from products or versions to streams gives you a set of many related child objects that you have to iterate over.
> 
> Using a ManyToManyField would simplify that approach and make the relationships look identical from both ends. That also makes this taxonomy identical to the others. Maybe there's reasons not to do so, though (other than the ones I outline above) - would appreciate any thoughts you have.

I think using ForeignKey is fine. It makes it explicit as to how you can link the different levels together. I think it's also fine to just have different sets of fields on each of the product levels without going through all the trouble to make it reusable for all child models. These will hardly ever change and I also don't see us adding yet another level to the product taxonomy.

Reply:
> That's odd. I don't see other abstract models having to define their child fields just to keep mypy happy. Looking at e.g. the Tag model, even if I comment out the `tagged_model` abstract method, mypy doesn't complain about `Tag` not having the `tagged_model` field defined.

It's a side effect of using ForeignKeys I think. We can't use an abstract property in the base class for e.g. the productvariants field. IIRC at least PyCharm / maybe other tools complain that subclasses aren't implementing all the required properties.

That productvariants field is there (reverse relation from the othermodel field on the ProductVariant model) but tools don't seem to detect it when it's part of a reverse relation. They only see it when it's explicitly defined as a ForeignKey (the forward side of the relation). I looked at defining all four taxonomy fields in the base class, but I think this interefered with the reverse relation's property getting set on the related model.

Reply:
Fixed - the newer version uses abstract properties on the ProductModel base class, then overrides these on each subclass to return either a manager or a class instance.

There are still type hints to help mypy figure things out, but the actual logic is closer to where it's being used and has less abstraction. Feel free to resolve this thread if it looks good, or please leave a comment if you think it can be improved further.

Reply:
Did you forget to push the actual changes? I don't see any new commits on my local branch.

Reply:
Latest version looks good!